src/hardware/vga_draw.cpp

```diff
@@ -469,57 +470,87 @@ static uint8_t* draw_linear_line_from_dac_palette(Bitu vidstart, Bitu /*line*/)
-		// This is used when the mouse cursor partially leaves the screen.
-		// It is arranged as bitmap of 16bits of bitA followed by 16bits of bitB, each
-		// AB bits corresponding to a cursor pixel. The whole map is 8kB in size.
-		memcpy(TempLine, &vga.mem.linear[ vidstart ], vga.draw.width);
-		// the index of the bit inside the cursor bitmap we start at:
-		Bitu sourceStartBit = ((lineat - vga.s3.hgc.originy) + vga.s3.hgc.posy)*64 + vga.s3.hgc.posx; 
-		// convert to video memory addr and bit index
-		// start adjusted to the pattern structure (thus shift address by 2 instead of 3)
-		// Need to get rid of the third bit, so "/8 *2" becomes ">> 2 & ~1"
-		Bitu cursorMemStart = ((sourceStartBit >> 2)& ~1) + (((uint32_t)vga.s3.hgc.startaddr) << 10);
-		Bitu cursorStartBit = sourceStartBit & 0x7;
-		// stay at the right position in the pattern
-		if (cursorMemStart & 0x2)
-			--cursorMemStart;
-		Bitu cursorMemEnd = cursorMemStart + ((64-vga.s3.hgc.posx) >> 2);
-		uint8_t* xat = &TempLine[vga.s3.hgc.originx]; // mouse data start pos. in scanline
-		for (Bitu m = cursorMemStart; m < cursorMemEnd;
-		     (m & 1) ? (m += 3) : ++m) {
-			// for each byte of cursor data
-			uint8_t bitsA = vga.mem.linear[m];
-			uint8_t bitsB = vga.mem.linear[m+2];
-			for (uint8_t bit=(0x80 >> cursorStartBit); bit != 0; bit >>= 1) {
-				// for each bit
-				cursorStartBit=0; // only the first byte has some bits cut off
-				if (bitsA&bit) {
-					if (bitsB&bit) *xat ^= 0xFF; // Invert screen data
-					//else Transparent
-				} else if (bitsB&bit) {
-					*xat = vga.s3.hgc.forestack[0]; // foreground color
-				} else {
-					*xat = vga.s3.hgc.backstack[0];
-				}
-				xat++;
-			}
+static uint8_t* draw_unwrapped_line_from_dac_palette_with_hwcursor(Bitu vidstart, Bitu /*line*/)
+{
+	// Draw the underlying line without the cursor
+	const auto line_addr = reinterpret_cast<uint32_t*>(
+	        draw_unwrapped_line_from_dac_palette(vidstart));
+
+	// Quick references to hardware cursor properties
+	const auto& cursor = vga.s3.hgc;
+	const auto line_at_y = (vidstart - (vga.config.real_start << 2)) / vga.draw.width;
+
+	// Is the mouse cursor pattern on this line?
+	if (cursor.posx >= vga.draw.width || line_at_y < cursor.originy ||
+	    line_at_y > (cursor.originy + (63U - cursor.posy))) {
+		return reinterpret_cast<uint8_t*>(line_addr);
+	}
+
+	// Draw mouse cursor
+	// ~~~~~~~~~~~~~~~~~
+	// The cursor is a 64x64 pattern which is shifted (inside the 64x64 mouse
+	// cursor space) to the right by posx pixels and up by posy pixels.
+	//
+	// This is used when the mouse cursor partially leaves the screen. It is
+	// arranged as bitmap of 16bits of bit A followed by 16bits of bit B, each AB
+	// bits corresponding to a cursor pixel. The whole map is 8kB in size.
+
+	// the index of the bit inside the cursor bitmap we start at:
+	const auto source_start_bit = ((line_at_y - cursor.originy) + cursor.posy) * 64 +
```

Comment:
Extract 64 to `pitch`?

Reply:
Sounds like a reasonable name for that constant. Will also see if it relates to the 63 (magic number) in here to, which are carried in from the original HW cursor function.